test/integration/src/test/java/io/pravega/test/integration/RestoreBackUpDataRecoveryTest.java

```diff
@@ -546,6 +582,428 @@ private void readEventsFromStreams(ClientFactoryImpl clientFactory, ReaderGroupM
+        assertTrue(watermark0.getLowerTimeBound() <= watermark0.getUpperTimeBound());
+        assertTrue(watermark1.getLowerTimeBound() <= watermark1.getUpperTimeBound());
+        assertTrue(watermark0.getLowerTimeBound() < watermark1.getLowerTimeBound());
+
+        Map<Segment, Long> positionMap0 = watermark0.getStreamCut().entrySet().stream().collect(
+                Collectors.toMap(x -> new Segment(SCOPE, STREAM1, x.getKey().getSegmentId()), Map.Entry::getValue));
+        Map<Segment, Long> positionMap1 = watermark1.getStreamCut().entrySet().stream().collect(
+                Collectors.toMap(x -> new Segment(SCOPE, STREAM1, x.getKey().getSegmentId()), Map.Entry::getValue));
+
+        StreamCut streamCutFirst = new StreamCutImpl(streamObj, positionMap0);
+        StreamCut streamCutSecond = new StreamCutImpl(streamObj, positionMap1);
+        Map<Stream, StreamCut> firstMarkStreamCut = Collections.singletonMap(streamObj, streamCutFirst);
+        Map<Stream, StreamCut> secondMarkStreamCut = Collections.singletonMap(streamObj, streamCutSecond);
+
+        return new ArrayList<Map<Stream, StreamCut>>(Arrays.asList(firstMarkStreamCut, secondMarkStreamCut));
+    }
+
+    private TimeWindow verifyEventsWithTimeBounds(Stream streamObj, EventStreamReader<Long> reader, EventRead<Long> event, TimeWindow currentTimeWindow) {
+        // read all events and verify that all events are below the bounds
+        while (event.getEvent() != null) {
+            Long time = event.getEvent();
+            log.info("timewindow = {} event = {}", currentTimeWindow, time);
+            assertTrue(currentTimeWindow.getLowerTimeBound() == null || time >= currentTimeWindow.getLowerTimeBound());
+            assertTrue(currentTimeWindow.getUpperTimeBound() == null || time <= currentTimeWindow.getUpperTimeBound());
+
+            TimeWindow nextTimeWindow = reader.getCurrentTimeWindow(streamObj);
+            assertTrue(currentTimeWindow.getLowerTimeBound() == null || nextTimeWindow.getLowerTimeBound() >= currentTimeWindow.getLowerTimeBound());
+            assertTrue(currentTimeWindow.getUpperTimeBound() == null || nextTimeWindow.getUpperTimeBound() >= currentTimeWindow.getUpperTimeBound());
+            currentTimeWindow = nextTimeWindow;
+
+            event = reader.readNextEvent(READ_TIMEOUT.toMillis());
+            if (event.isCheckpoint()) {
+                event = reader.readNextEvent(READ_TIMEOUT.toMillis());
+            }
+        }
+        return currentTimeWindow;
+    }
+
+    /**
+     * Adds water marks to the watermarks queue.
+     */
+    private void fetchWatermarks(RevisionedStreamClient<Watermark> watermarkReader, LinkedBlockingQueue<Watermark> watermarks,
+                                 AtomicBoolean stop) {
+        AtomicReference<Revision> revision = new AtomicReference<>(watermarkReader.fetchOldestRevision());
+
+        Futures.loop(() -> !stop.get(), () -> Futures.delayedTask(() -> {
```

Comment:
One comment here. I see you stop this loop by setting `stop` too true and then you immediately move on to the next thing in your test.

You never wait on this Future  loop to complete. You may set stop to true, but if it's currently waiting (for 10s), it will still execute the subsequent code eventually which could result in your `watermarks` collection to be modified. That will most likely cause issues.

Consider rewriting this so that this method:
1. Returns the `Futures.loop`
2. The callsite holds it in a variable and waits for it to end.
3. Reducing the delay between iterations to 1s. That will prevent this from executing long after your test is done.
4. In the callback (line 933), bail out if `stop` is false.

Reply:
1. Returned Futures.loop
2. The callsite holds it in a variable and waits for it to end.
3. Reduced the delay to 1s.
4. I didn't get this one.

Reply:
For #4, if you set `stop` to true, but are still waiting for that delayed future, the callback will still execute, even though `stop` is true. Ideally it shouldn't.

Reply:
Exiting from delayed future if stop is set to `true`.